v0.2.1/apps/chrome-extension/data-processing.js

```diff
@@ -8,14 +8,38 @@
 /**
  * Reduce the deeply nested structure of url=>entities-of-many-types to a
  * flatter, combined form of type=>entities.
+ *
+ * For example, the input would be of the form
+ * {'http://g.co': [{@type: 'typeA', 'a': 1}},
+ * the output would be
+ * {typeA: [{@type: 'typeA', 'a': 1}]}.
  */
 function flatten(entities) {
-  return Object.entries(entities).reduce( (accumulator, [key, value]) => {
+  return Object.entries(entities).reduce((accumulator, [key, value]) => {
     value.forEach(entry => {
-      // TODO(smalls) need to dedup in here as well
       let type = entry['@type'];
-      accumulator[type] ? accumulator[type].push(entry) : accumulator[type] = [entry];
+      accumulator[type]
+        ? accumulator[type].push(entry)
+        : (accumulator[type] = [entry]);
     });
     return accumulator;
   }, new Object());
 }
+
+/**
+ * Removes duplicate entries. Expects the input to match the output format of
+ * #flatten().
+ */
+function deduplicate(entities) {
+  return Object.entries(entities).reduce((accumulator, [key, values]) => {
+    accumulator[key] = values.reduce((accumulator, value) => {
+      let isIncluded = accumulator.reduce(
+        (a, av) => _.isEqual(av, value) || a,
```

Comment:
I imported lodash for the deep equals method - _.isEqual() in the line above. Does anyone know of a better way? Is there a preferred library?

@sjmiles